apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/RdaVersion.java

```diff
@@ -0,0 +1,181 @@
+   *
+   * @return A {@link RdaVersionBuilder}.
+   */
+  public static RdaVersionBuilder builder() {
+    return new RdaVersionBuilder();
+  }
+
+  /**
+   * Determines if this {@link RdaVersion} specifies a version or range that would accept the given
+   * {@link RdaVersion}.
+   *
+   * @param rdaVersionString The version being checked to see if it falls within range.
+   * @return True if the given {@link RdaVersion} is within range of this instance's {@link
+   *     RdaVersion} requirement.
+   */
+  public boolean allows(String rdaVersionString) {
+    // When we connect to S3 or some other source, there is a prefix on the version to distinguish
+    // it.
+    String[] bits = rdaVersionString.split(":");
+    String versionString = bits[bits.length - 1];
+    // Parse the version
+    RdaVersion rdaVersion = builder().versionString(versionString).build();
+
+    boolean compatable;
+
+    switch (compatability) {
+      case MAJOR:
+        compatable = rdaVersion.major == major;
+        break;
+      case MINOR:
+        compatable = rdaVersion.major == major && rdaVersion.minor == minor;
+        break;
+      default:
+        compatable =
+            rdaVersion.major == major && rdaVersion.minor == minor && rdaVersion.patch == patch;
+    }
+
+    return compatable;
+  }
+
+  /** The potential compatability levels that can be specified. */
+  public enum CompatabilityLevel {
+    /** Must be same major version. */
+    MAJOR,
+    /** Must be same Major/Minor version. */
+    MINOR,
+    /** Must match exact same Major, Minor, and Patch version. */
+    PATCH
+  }
+
+  /** Helper class for building {@link RdaVersion} objects. */
+  public static class RdaVersionBuilder {
+
+    /** The version string to parse. */
+    private String versionString;
+
+    /** Empty private constructor to prevent direct instantiation. */
+    private RdaVersionBuilder() {
+      // Nothing to do
+    }
+
+    /**
+     * Sets the version string to parse when building the {@link RdaVersion} object.
+     *
+     * @param versionString The version string to set.
+     * @return A reference to this builder instance.
+     */
+    public RdaVersionBuilder versionString(String versionString) {
+      this.versionString = versionString;
+      return this;
+    }
+
+    /**
+     * Builds an {@link RdaVersion} object instance using the previously given parameters.
+     *
+     * @return An {@link RdaVersion} object built with the previously provided parameters.
+     */
+    public RdaVersion build() {
+      if (versionString != null) {
+        CompatabilityLevel level = CompatabilityLevel.PATCH;
+
+        // Array includes: { Level, Major, Minor, Patch }
+        Integer[] versionParts = new Integer[] {0, 0, 0, 0};
+        int stage = 0;
+
+        // Go through each character and parse out the version parts
+        for (int i = 0; i < versionString.length() && stage < versionParts.length; ) {
+          char c = versionString.charAt(i);
+
```

Comment:
I think a regex would be easier to understand here.  Something like 

```
var regex = Pattern.compile("(?<type>[\\^~]?)(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?");
var matcher = regex.matcher(whatever);
if (!matcher.matches()) {
  throw ...
}
```
Then you get the values right from the matched groups.